wire/decoder.go

```diff
@@ -0,0 +1,367 @@
+package wire
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"math"
+	"net"
+	"time"
+)
+
+//integer sizes
+const (
+	ByteSize    = 1
+	ShortSize   = 2
+	IntegerSize = 4
+	LongSize    = 8
+)
+
+// ConnInfo contains information about the database connection. This is returned
+// from the database on a successful login.
+type ConnInfo struct {
+
+	// Host is the host ID of the volt node
+	Host int32
+
+	// Connection unique id for the connection in the current volt node
+	Connection int64
+
+	// Time  in which the cluster was started
+	ClusterStart time.Time
+
+	// IPV4 address of the leader node
+	LeaderAddr struct {
+		Value int32
+		IP    net.IP
+	}
+
+	// Representation of the build of the connected node
+	Build string
+}
+
+// Decoder is voltdb wire protocol decoder
+type Decoder struct {
+	r io.Reader
+}
+
+// NewDecoder returns a new Decoder that decods values read from src
+func NewDecoder(src io.Reader) *Decoder {
+	return &Decoder{r: src}
+}
+
+// SetReader replaces the underlying reader, next call to Decode methods will
+// read from this
+func (d *Decoder) SetReader(r io.Reader) {
```

Comment:
I think the whole point of the new decoder API is to help improve performance, keeping that in mind making this reusable is the best logical choice. `Encoder` is reusable see `(*Encoder).Reset`